### PR TITLE
Seprate network interfaces unit test

### PR DESCRIPTION
**Pull request by assafad**
### What this PR does
Before this PR:
Network interfaces unit test was using a single function that gets one of three possible inputs. Those inputs require different implementation in most of the function, so the function had many conditions, for every possible input. This is not readable and maintainable.
After this PR:
Instead, separate to 3 tests, and use `beforeEach()` for the common code.

/sig code-quality
/cc @EdDev

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (EdDev) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:417]: Perhaps use this: `When("Kubevirt has network-interface defined", ...`

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (EdDev) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:430]: Perhaps replace `add` with `use`. Same for the others.

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (EdDev) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:447]: Maybe you can use `pointer.P(true)` instead.

**Review by EdDev — COMMENTED** [review]
Thank you!
Some small comments inline.

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (assafad) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go]: Done

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (assafad) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go]: Done

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (assafad) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go]: Done

**Review by EdDev — APPROVED** [review]

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (iholder101) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:425]: Isn't this part redandunt?

**Review by iholder101 — COMMENTED** [review]
Was this alternative considered?

```go
DescribeTable("should add the default network interface",
	func(interfaceName string) {
		testutils.UpdateFakeKubeVirtClusterConfig(kvInformer, &v1.KubeVirt{
			Spec: v1.KubeVirtSpec{
				Configuration: v1.KubeVirtConfiguration{
					NetworkConfiguration: &v1.NetworkConfiguration{
						NetworkInterface:     interfaceName,
						PermitSlirpInterface: k6tpointer.P(true),
					},
				},
			},
		})

		_, vmiSpec, _ := getMetaSpecStatusFromAdmit(rt.GOARCH)
		vmiInterface := vmiSpec.Domain.Devices.Interfaces[0]
		switch interfaceName {
		case "bridge":
			Expect(vmiInterface.Bridge).NotTo(BeNil())
		case "masquerade":
			Expect(vmiInterface.Masquerade).NotTo(BeNil())
		case "slirp":
			Expect(vmiInterface.Slirp).NotTo(BeNil())
		}
	},
	Entry("as bridge", "bridge"),
	Entry("as masquerade", "masquerade"),
	Entry("as slirp", "slirp"),
)
```

AFAICT the `expectedIface` variable is completely redundant. If we remove this part the configuration remains the same for every case, and the only difference is the one-line expectation which IMO fits the switch-case pattern. WDYT?

**Comment by assafad** [comment]
@iholder101 I don't like the following part:
```go
switch interfaceName {
case "bridge":
	Expect(vmiInterface.Bridge).NotTo(BeNil())
case "masquerade":
	Expect(vmiInterface.Masquerade).NotTo(BeNil())
case "slirp":
	Expect(vmiInterface.Slirp).NotTo(BeNil())
}
Entry("as bridge", "bridge"),
Entry("as masquerade", "masquerade"),
Entry("as slirp", "slirp"),
```
where the function checks for every possible input. In such cases I prefer to see 3 different tests. @EdDev wdyt?

**Comment by EdDev** [comment]
> @iholder101 I don't like the following part:
> 
> ```go
> switch interfaceName {
> case "bridge":
> 	Expect(vmiInterface.Bridge).NotTo(BeNil())
> case "masquerade":
> 	Expect(vmiInterface.Masquerade).NotTo(BeNil())
> case "slirp":
> 	Expect(vmiInterface.Slirp).NotTo(BeNil())
> }
> Entry("as bridge", "bridge"),
> Entry("as masquerade", "masquerade"),
> Entry("as slirp", "slirp"),
> ```
> 
> where the function checks for every possible input. In such cases I prefer to see 3 different tests. @EdDev wdyt?

Having conditions in the common function on the input is an anti-pattern.
It implies you need 3 different tests.

**Comment by xpivarc** [comment]
/hold
What about something like https://github.com/kubevirt/kubevirt/pull/11783/commits/658e3505ee3dac74223d4921144d647c4e6dba75#diff-6a5a8576ebb4952ea056750d75d85e96d53eb62406defce2df9091dc9c0c0c2dL5464-R5475 ?
I am not sure I would suggest removing each table test.

**Comment by assafad** [comment]
> /hold What about something like [658e350#diff-6a5a8576ebb4952ea056750d75d85e96d53eb62406defce2df9091dc9c0c0c2dL5464-R5475](https://github.com/kubevirt/kubevirt/commit/658e3505ee3dac74223d4921144d647c4e6dba75#diff-6a5a8576ebb4952ea056750d75d85e96d53eb62406defce2df9091dc9c0c0c2dL5464-R5475) ? I am not sure I would suggest removing each table test.

Not sure what you mean, you linked to a deleted code which I think is not from main.

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (xpivarc) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:417]:
```
DescribeTable("should add the default network interface",
		func(iface string, field gstruct.Field) {
			permit := true
			testutils.UpdateFakeKubeVirtClusterConfig(kvInformer, &v1.KubeVirt{
				Spec: v1.KubeVirtSpec{
					Configuration: v1.KubeVirtConfiguration{
						NetworkConfiguration: &v1.NetworkConfiguration{
							NetworkInterface:     expectedIface,
							PermitSlirpInterface: &permit,
						},
					},
				},
			})

			_, vmiSpec, _ := getMetaSpecStatusFromAdmit(rt.GOARCH)
                       Expect(vmiSpec.Domain.Devices.Interfaces[0]).To(gstruct.MatchFields(gstruct.IgnoreExtras,
					gstruct.Fields{
						"InterfaceBindingMethod": gstruct.MatchFields(gstruct.IgnoreExtras, field),
					},
				))
			
		},
		Entry("as bridge", "bridge", gstruct.Fields{"Bridge": Not(BeNil())}),
		Entry("as masquerade", "masquerade", , gstruct.Fields{"Masquerade": Not(BeNil())}),
		Entry("as slirp", "slirp" , gstruct.Fields{"Slirp": Not(BeNil())}),
	)
```

**Comment by xpivarc** [comment]
> > /hold What about something like [658e350#diff-6a5a8576ebb4952ea056750d75d85e96d53eb62406defce2df9091dc9c0c0c2dL5464-R5475](https://github.com/kubevirt/kubevirt/commit/658e3505ee3dac74223d4921144d647c4e6dba75#diff-6a5a8576ebb4952ea056750d75d85e96d53eb62406defce2df9091dc9c0c0c2dL5464-R5475) ? I am not sure I would suggest removing each table test.
> 
> Not sure what you mean, you linked to a deleted code which I think is not from main.

Please have a look at https://github.com/kubevirt/kubevirt/pull/11713/files#r1579004437

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (EdDev) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:417]: This is interesting.
But it is IMO hard to read.

I'm supposed to be able and understand the test quickly, but here the cognitive effort needed for me to interpret this is high.

Comment (assafad) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:417]: I agree with @EdDev on this. 
This is cool, but much less readable IMO. Also, with your implementation we permit slirp interface also in other interfaces tests, which IIUC is not needed and is only confusing.

Comment (xpivarc) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:417]: > This is interesting. But it is IMO hard to read.
> 
> I'm supposed to be able and understand the test quickly, but here the cognitive effort needed for me to interpret this is high.

What is the cognitive load here?

**Comment by EdDev** [comment]
> /hold What about something like [658e350#diff-6a5a8576ebb4952ea056750d75d85e96d53eb62406defce2df9091dc9c0c0c2dL5464-R5475](https://github.com/kubevirt/kubevirt/commit/658e3505ee3dac74223d4921144d647c4e6dba75#diff-6a5a8576ebb4952ea056750d75d85e96d53eb62406defce2df9091dc9c0c0c2dL5464-R5475) ? I am not sure I would suggest removing each table test.

@xpivarc 
The tests are asserting different fields and types and therefore IMO do not fit well into a test-table.
The suggested alternatives looks like a hack to me and is not readable.

Please consider to unhold this change. It is still better than the previous version and we have a maintainer agreeing to support it.

**Comment by AlonaKaplan** [comment]
/approve

> /hold What about something like [658e350#diff-6a5a8576ebb4952ea056750d75d85e96d53eb62406defce2df9091dc9c0c0c2dL5464-R5475](https://github.com/kubevirt/kubevirt/commit/658e3505ee3dac74223d4921144d647c4e6dba75#diff-6a5a8576ebb4952ea056750d75d85e96d53eb62406defce2df9091dc9c0c0c2dL5464-R5475) ? I am not sure I would suggest removing each table test.

I find the PR suggestion more readable than this option.

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go
Comment (xpivarc) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:417]: I am still not sure what is hard about this... I think it reads pretty well, match fields, field InterfaceBindingMethod needs to match fields based on parameter...

Comment (EdDev) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:417]: I just realized, that there is a straightforward option here that keeps using the table.
@assafad , as the PR got an automatic hold due to too many retries, perhaps it is a good opportunity to consider it (if it really works).

```go
DescribeTable("should add the default network interface", func(defaultBindingMethod string, expectedIfaceBindingMethod v1.InterfaceBindingMethod) {
	testutils.UpdateFakeKubeVirtClusterConfig(kvInformer, &v1.KubeVirt{
		Spec: v1.KubeVirtSpec{
			Configuration: v1.KubeVirtConfiguration{
				NetworkConfiguration: &v1.NetworkConfiguration{
					NetworkInterface:     expectedIface,
					PermitSlirpInterface: pointer.P(true),
				},
			},
		},
	})
	_, vmiSpec, _ := getMetaSpecStatusFromAdmit(rt.GOARCH)
	Expect(vmiSpec.Domain.Devices.Interfaces[0].InterfaceBindingMethod).To(Equal(expectedIfaceBindingMethod))
},
    Entry("as bridge", "bridge", v1.InterfaceBindingMethod{Bridge: &v1.InterfaceBridge{}}),
    Entry("as masqurade", "masquerade", v1.InterfaceBindingMethod{Masquerade: &v1.InterfaceMasquerade{}}),
    Entry("as slirp", "slirp", v1.InterfaceBindingMethod{Slirp: &v1.InterfaceSlirp{}}),
)
```

Comment (assafad) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:417]: Thanks, it works. Done.

**Review by EdDev — APPROVED** [review]
Very nice, thanks!

**Comment by brianmcarey** [comment]
/override pull-kubevirt-fossa 

fossa is failing on the main branch with the same issue - not related to this PR.

**Pull request merged**
